File: src/models/order.js
const { Sequelize, DataTypes } = require('sequelize');
const config = require('../core/config');

const sequelize = new Sequelize(
  config.database.name,
  config.database.user,
  config.database.password,
  {
    host: config.database.host,
    port: config.database.port,
    dialect: 'postgres',
    logging: false
  }
);

const Order = sequelize.define('Order', {
  id: {
    type: DataTypes.UUID,
    primaryKey: true,
    defaultValue: DataTypes.UUIDV4
  },
  pieType: {
    type: DataTypes.STRING(50),
    allowNull: false,
    field: 'pie_type'
  },
  customerName: {
    type: DataTypes.STRING(255),
    allowNull: false,
    field: 'customer_name'
  },
  customerEmail: {
    type: DataTypes.STRING(255),
    allowNull: false,
    field: 'customer_email'
  },
  customerPhone: {
    type: DataTypes.STRING(20),
    field: 'customer_phone'
  },
  deliveryStreet: {
    type: DataTypes.STRING(255),
    allowNull: false,
    field: 'delivery_street'
  },
  deliveryCity: {
    type: DataTypes.STRING(100),
    allowNull: false,
    field: 'delivery_city'
  },
  deliveryState: {
    type: DataTypes.STRING(2),
    allowNull: false,
    field: 'delivery_state'
  },
  deliveryZip: {
    type: DataTypes.STRING(10),
    allowNull: false,
    field: 'delivery_zip'
  },
  currentState: {
    type: DataTypes.STRING(20),
    allowNull: false,
    field: 'current_state'
  },
  pickerJobId: {
    type: DataTypes.STRING(255),
    field: 'picker_job_id'
  },
  bakerJobId: {
    type: DataTypes.STRING(255),
    field: 'baker_job_id'
  },
  deliveryId: {
    type: DataTypes.STRING(255),
    field: 'delivery_id'
  },
  estimatedDelivery: {
    type: DataTypes.DATE,
    field: 'estimated_delivery'
  }
}, {
  tableName: 'orders',
  timestamps: true,
  underscored: true
});

const StateHistory = sequelize.define('StateHistory', {
  id: {
    type: DataTypes.INTEGER,
    primaryKey: true,
    autoIncrement: true
  },
  orderId: {
    type: DataTypes.UUID,
    allowNull: false,
    field: 'order_id'
  },
  fromState: {
    type: DataTypes.STRING(20),
    field: 'from_state'
  },
  toState: {
    type: DataTypes.STRING(20),
    allowNull: false,
    field: 'to_state'
  },
  timestamp: {
    type: DataTypes.DATE,
    defaultValue: DataTypes.NOW
  },
  notes: {
    type: DataTypes.TEXT
  },
  errorMessage: {
    type: DataTypes.TEXT,
    field: 'error_message'
  }
}, {
  tableName: 'state_history',
  timestamps: false
});

Order.hasMany(StateHistory, { foreignKey: 'orderId', as: 'history' });
StateHistory.belongsTo(Order, { foreignKey: 'orderId' });

module.exports = {
  sequelize,
  Order,
  StateHistory
};

File: src/models/order.py
import uuid
from datetime import datetime
from typing import Optional
from sqlalchemy import String, DateTime, Text, ForeignKey, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship
from src.database import Base
from src.core.state_machine import OrderState


class Order(Base):
    __tablename__ = "orders"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: f"ord_{uuid.uuid4().hex[:8]}")
    pie_type: Mapped[str] = mapped_column(String(50), nullable=False)
    customer_name: Mapped[str] = mapped_column(String(255), nullable=False)
    customer_email: Mapped[str] = mapped_column(String(255), nullable=False)
    customer_phone: Mapped[Optional[str]] = mapped_column(String(20), nullable=True)
    delivery_street: Mapped[str] = mapped_column(String(255), nullable=False)
    delivery_city: Mapped[str] = mapped_column(String(100), nullable=False)
    delivery_state: Mapped[str] = mapped_column(String(2), nullable=False)
    delivery_zip: Mapped[str] = mapped_column(String(10), nullable=False)
    current_state: Mapped[str] = mapped_column(String(20), default=OrderState.ORDERED.value)
    picker_job_id: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    baker_job_id: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    delivery_id: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    estimated_delivery: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    error_message: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), onupdate=func.now())

    history: Mapped[list["StateHistory"]] = relationship("StateHistory", back_populates="order", lazy="selectin")


class StateHistory(Base):
    __tablename__ = "state_history"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    order_id: Mapped[str] = mapped_column(String(36), ForeignKey("orders.id"), nullable=False)
    from_state: Mapped[Optional[str]] = mapped_column(String(20), nullable=True)
    to_state: Mapped[str] = mapped_column(String(20), nullable=False)
    timestamp: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    notes: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    error_message: Mapped[Optional[str]] = mapped_column(Text, nullable=True)

    order: Mapped["Order"] = relationship("Order", back_populates="history")
